rpcs3/util/vm_native.cpp

```diff
@@ -87,6 +89,100 @@ namespace utils
 #ifdef _WIN32
 	DYNAMIC_IMPORT("KernelBase.dll", VirtualAlloc2, PVOID(HANDLE Process, PVOID Base, SIZE_T Size, ULONG AllocType, ULONG Prot, MEM_EXTENDED_PARAMETER*, ULONG));
 	DYNAMIC_IMPORT("KernelBase.dll", MapViewOfFile3, PVOID(HANDLE Handle, HANDLE Process, PVOID Base, ULONG64 Off, SIZE_T ViewSize, ULONG AllocType, ULONG Prot, MEM_EXTENDED_PARAMETER*, ULONG));
+	DYNAMIC_IMPORT("KernelBase.dll", UnmapViewOfFile2, BOOL(HANDLE Process, PVOID BaseAddress, ULONG UnmapFlags));
+
+	const bool has_win10_memory_mapping_api()
+	{
+		return VirtualAlloc2 && MapViewOfFile3 && UnmapViewOfFile2;
+	}
+
+	struct map_info_t
+	{
+		u64 addr = 0;
+		u64 size = 0;
+		atomic_t<u8> state{};
+	};
+
+	lf_array<map_info_t, 32> s_is_mapping{};
+
+	bool is_memory_mappping_memory(u64 addr)
+	{
+		if (!addr)
+		{
+			return false;
+		}
+
+		const u64 map_size = s_is_mapping.size();
+
+		for (u64 i = map_size - 1; i != umax; i--)
+		{
+			const auto& info = s_is_mapping[i];
+
+			if (info.state == 1)
+			{
+				if (addr >= info.addr && addr < info.addr + info.size)
+				{
+					return true;
+				}
+			}
+		}
+
+		return false;
+	}
+
+	u64 unmap_mappping_memory(u64 addr, u64 size)
+	{
+		if (!addr || !size)
+		{
+			return false;
+		}
+
+		const u64 map_size = s_is_mapping.size();
+
+		for (u64 i = map_size; i != umax; i--)
+		{
+			auto& info = s_is_mapping[i];
+
+			if (info.state == 1)
+			{
+				if (addr == info.addr && size == info.size)
+				{
+					if (info.state.compare_and_swap_test(1, 0))
+					{
+						return info.size;
+					}
+				}
+			}
+		}
+
+		return false;
+	}
+
+	bool map_mappping_memory(u64 addr, u64 size)
+	{
+		if (!addr || !size)
+		{
+			return false;
+		}
+
+		for (u64 i = 0;; i++)
+		{
+			auto& info = s_is_mapping[i];
+
+			if (!info.addr && info.state.compare_and_swap_test(0, 2))
+			{
+				info.addr = addr;
+				info.size = size;
+				info.state = 1;
+				return true;
+			}
+		}
```

Comment:
Theoretically this misses a definite exit condition.
Is it guaranteed to always find a free spot?

Reply:
Yes, when the array buffer is all checked, it extends the array with data that must be available for this function.